app/services/course/material/download_service.rb

```diff
@@ -0,0 +1,64 @@
+class Course::Material::DownloadService
+  require 'zip'
+  require 'uri'
+  require 'open-uri'
+
+  class << self
+    # Download the contents and zip them.
+    #
+    # @param [Array<Course::Material|Course::Material::Folder>] contents The contents to be
+    #   downloaded.
+    # @return [String] The url to the zip file.
+    def download_and_zip(contents)
```

Comment:
your job can reconstruct the folder hierarchy given just the list of materials =.= the material has a reference to its folder, which has the reference to its parent all the way until folder A right?